lib/Analysis/AxisInfo.cpp

```diff
@@ -882,9 +882,7 @@ class ShLIOpAxisInfoVisitor final : public BinaryOpVisitorImpl<arith::ShLIOp> {
 
   int64_t getDivisibility(arith::ShLIOp op, const AxisInfo &lhs,
                           const AxisInfo &rhs, int dim) override {
-    auto shift = rhs.getConstantValue().has_value()
-                     ? rhs.getConstantValue().value()
-                     : rhs.getDivisibility(dim);
+    auto shift = rhs.getConstantValue().value_or(0);
```

Comment:
FYI, 0's divisibility is usually the maximum possible power of 2 integer of its type. Maybe we could do special handling here.

Reply:
> For example, if you have %a << %b, where both a and b are unknown values but with divisibility 4 and 4, then we could estimate the divisibility of the result as 4 << 4.

I don't understand how that is correct. if %b is 0 then this is not true?

>FYI, 0's divisibility is usually the maximum possible power of 2 integer of its type. Maybe we could do special handling here.

I think we would have to fix a bunch of other cases. For instance when we calculate the divisibility of an unknown %a in %a << 1 we will assume divisibility by 2 however the value may still be 0.

Reply:
> As I mentioned, I'm worried about its a bit too conservative since rhs is usually not 0.

I agree, I couldn't find a better way that would always be correct

Reply:
If I've understood well the definition of divisibility, you want to compute the largest power of two that divides `a << b`.
If there's no overflow, this is the same as `a * 2^b`.
```c++
int64_t log_div_lhs;
if (lhs.getConstantValue().has_value()) {
    auto lhs_v = *lhs.getConstantValue();
    // Or implement ctz manually or smth
    log_div_lhs = lhs_v ? __builtin_ctz(lhs_v) : 64;
} else {
    log_div_lhs = log2Int(lhs.getDivisibility(dim));
}
int64_t log_div_rhs;
if (rhs.getConstantValue().has_value()) {
    log_div_rhs = *rhs.getConstantValue();
} else {
    log_div_rhs = rhs.getDivisibility(dim);
}
auto shift = int64_t{1} << (log_div_lhs + log_div_rhs);
```
or something along this lines (not tested)

Reply:
> If I've understood well the definition of divisibility, you want to compute the largest power of two that divides a << b.

Correct

> If there's no overflow, this is the same as a * 2^b.

yes

> log_div_lhs = log2Int(lhs.getDivisibility(dim));

this assumes that ` lhs >= lhs.getDivisibility(dim)` which I think is not always true as lhs can be 0.

Reply:
In that case, if `lhs == 0`, then I chose `1 << 64` as its divisibility, as @Jokeren proposed in the branch above that. If we have that, then this code would still be correct... but I guess we don't have it?

Reply:
if lhs == 0 then `lhsDivisibility` should already be MAX_INT, so I believe this case should already work.